sql: a txn respects the ModificationTime of a table descriptor

vivekmenezes opened the pull request:
The first commit is in another PR.

vivekmenezes committed:
sql: a txn respects the ModificationTime of a table descriptor

A sql transaction is supposed to respect the validity window
of any table descriptor it uses:
table.ModificationTime <= txn.Timestamp < table.expiration.
The txn would always respect the expiration time of a table
version. This change enforced the lower ModificationTime bound.

This change allows a transaction that was started
before a schema change to use the schema before the
schema change. The transaction respects the ModificationTime
of a table descriptor and requests an earlier version of a
table descriptor. Reads are allowed using an older descriptor
while writes using the older descriptor get pushed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16842)
<!-- Reviewable:end -->

vivekmenezes commented:
@andreimatei I've rebased this change. This is one of two PR I'm working on. This one ensures that we use the correct version of a table descriptor in a transaction and therefore don't have to apply any special retries in case a transaction is using an old timestamp. I'm working on another PR which is the final one to acquire leases without using the users transaction. These are the two PRs I want to get merged by the end of week for 1.1. Thanks

andreimatei commented:
Review status: 0 of 7 files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease.go, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuNAaH4Vcmp0NCv9EM:-KpuNAaH4Vcmp0NCv9EN:bvbkv3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L95)):*
> ```Go
> }
> 
> // hasSomeLifeLeft returns true if the lease has at least a minimum of
> ```

the comment talks about a lease, but we're operating on a `tableVersionState` that doesn't necessarily correspond to a lease (right?). Should this check `s.leased` and error? Or, if this is supposed to work for non-lease descriptors, should we stop using `MinLeaseDuration` for those and rename the method to `requiresRefresh`?

Also, the "can be used" part of the comment is vestigial, isn't it?

---

*[pkg/sql/lease.go, line 359 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuOX_HUmYGZMgAKdO2:-KpuOX_HUmYGZMgAKdO3:b-kzxw2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L359)):*
> ```Go
> 
> 			tableDesc.Version++
> 			now := txn.OrigTimestamp()
> ```

s/now/modTime

And I think we should use `txn.Timestamp()`. In case of serializable txns (which this is), they'll be the same, but `Timestamp` more is what we generally use to indicate the timestamp at which a txn will commit.

---

*[pkg/sql/lease.go, line 529 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuPMskd5mk2AChjA3h:-KpuPMskd5mk2AChjA3i:b-k2zezc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L529)):*
> ```Go
> 	mu syncutil.Mutex
> 	// The active leases for the table: sorted by their version.
> 	// There may be more than one active lease when the system is
> ```

this comment should be updated I think

---

*[pkg/sql/lease.go, line 547 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuQeH3hp7h_4DLr-K8:-KpuQeH3hp7h_4DLr-K9:basd8oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L547)):*
> ```Go
> // acquire returns the latest version of a table. The table will have its
> // refcount incremented, so the caller is responsible for calling release() on it.
> func (t *tableState) acquire(
> ```

I'm sorry, I don't understand this function. It's supposed to return "the latest version" but it uses a passed-in txn, and sometimes it doesn't go to the database at all. Is the "latest" part just wrong?

---

*[pkg/sql/lease.go, line 552 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuPRXDrdPJVSS0rbKZ:-KpuPRXDrdPJVSS0rbK_:b-sqolj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L552)):*
> ```Go
> 	t.mu.Lock()
> 	defer t.mu.Unlock()
> 	timestamp := txn.OrigTimestamp()
> ```

here `OrigTimestamp` is good (depending on what exactly this method is supposed to do, which I'm not sure), as we probably care about the the transaction's read timestamp

---

*[pkg/sql/lease.go, line 554 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuSWyVCO85GrMxrDmX:-KpuSWyVCO85GrMxrDmY:bsdtfnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L554)):*
> ```Go
> 	timestamp := txn.OrigTimestamp()
> 	for {
> 		if s := t.active.findNewest(); s != nil && s.hasSomeLifeLeft(timestamp) {
> ```

as I was commenting on `hasSomeLifeLeft()`, I'm not sure what it's supposed to do, but I suspect that this will have to change: I think there has to be a case where we refresh a lease async if it's about to expire, but we also go ahead and use it

---

*[pkg/sql/lease.go, line 560 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv-7wSmtugH2cJFKHF:-Kpv-7wSmtugH2cJFKHG:b-f7ymxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L560)):*
> ```Go
> 			}
> 			// The txn cannot use the latest valid table descriptor.
> 			return t.findForTimestamp(ctx, timestamp, m)
> ```

if we have this `findForTimestamp

---

*[pkg/sql/lease.go, line 591 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvWGEU9QhTHWFe3z3f:-KpvWGEVtRG2KMTa6enS:bp0hheu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L591)):*
> ```Go
> 
> // Find the table descriptor valid for the particular timestamp. If the valid
> // version doesn't exist it is read from the store.
> ```

please comment that the refcount is incremented

---

*[pkg/sql/lease.go, line 592 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvarPiF5KTD41LW9fa:-KpvarPiF5KTD41LW9fb:b-fwrbdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L592)):*
> ```Go
> // Find the table descriptor valid for the particular timestamp. If the valid
> // version doesn't exist it is read from the store.
> func (t *tableState) findForTimestamp(
> ```

so this doesn't lease, right? Please explain what assumption this has - that a more recent version than timestamp exists in `t.active`?

---

*[pkg/sql/lease.go, line 598 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv0CMmH5scpAnXX6zB:-Kpv0CMmH5scpAnXX6zC:b-7fl0u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L598)):*
> ```Go
> 	// unsuccessful run reads a table descriptor version from the store
> 	// and inserts it.
> 	for {
> ```

I don't think we need the loop... Once you go to the store, you're guaranteed to get the right descriptor, aren't you?

---

*[pkg/sql/lease.go, line 601 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvaHGU5teOirkmtdds:-KpvaHGU5teOirkmtddt:betago5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L601)):*
> ```Go
> 		var expiration hlc.Timestamp
> 		for i := len(t.active.data) - 1; i >= 0; i-- {
> 			if table := t.active.data[i]; !timestamp.Less(table.ModificationTime) {
> ```

I found this hard to understand. The `timestamp.Less` are hard to read. I think an ascii diagram might help, for example.

---

*[pkg/sql/lease.go, line 614 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv_LKmZsejaH1FhWpF:-Kpv_LKmZsejaH1FhWpG:boqstko) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L614)):*
> ```Go
> 					return nil, fmt.Errorf("requesting a table version ahead of latest version")
> 				}
> 				expiration = t.active.data[next].ModificationTime
> ```

if this is exclusive, let's please document that where `expiration` is defined

---

*[pkg/sql/lease.go, line 639 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv_UqjRae_mAcpUah0:-Kpv_UqjRae_mAcpUah1:b-7n7htk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L639)):*
> ```Go
> }
> 
> // Get the table descriptor valid for the expiration time from the store.
> ```

"valid for the expiration time" sounds very weird. Let's please explain it some more. 

---

*[pkg/sql/lease.go, line 640 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv2W1vCqkDJKv_1_Z-:-Kpv2W1vCqkDJKv_1_Z0:b-5h1mbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L640)):*
> ```Go
> 
> // Get the table descriptor valid for the expiration time from the store.
> func (t *tableState) getForExpiration(
> ```

I think this should be a method on the `LeaseStore` (which should be renamed) and return a naked descriptor, otherwise it's awkward that this uses the `LeaseStore.db`. And also, if it stays on `tableState`, I think it's weird that it doesn't increment the refcount.

---

*[pkg/sql/lease.go, line 640 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvVgp9RIHfi7iH0uGq:-KpvVgp9RIHfi7iH0uGr:bq2osrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L640)):*
> ```Go
> 
> // Get the table descriptor valid for the expiration time from the store.
> func (t *tableState) getForExpiration(
> ```

it'd be cool to document what happens if you ask for an expiration beyond the last GC. I had to fish for the answer for a while, and it is a naked error in `replica.requestCanProceed()`. It'd be even cooler if we gave that error an error code.

---

*[pkg/sql/lease.go, line 648 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv3PIRaS03LHnM7PG6:-Kpv3PIRaS03LHnM7PG7:b-pxsf2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L648)):*
> ```Go
> 		prevTimestamp := expiration
> 		prevTimestamp.WallTime--
> 		txn.SetFixedTimestamp(prevTimestamp)
> ```

it'd be :100:  if there was a virtual table for descriptors and we'd do this with a sql time-travel SELECT :)

---

*[pkg/sql/lease.go, line 751 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvWs6LFDT2ly_su26V:-KpvWs6LFDT2ly_su26W:b-d8ti5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L751)):*
> ```Go
> }
> 
> // releaseInactiveVersions removed inactive versions in t.active.data with refcount 0.
> ```

removed-removes

Maybe `removeInactive...` is a better name.

---

*[pkg/sql/lease.go, line 1094 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv2CGNwcnUfsztCaOF:-Kpv2CGNwcnUfsztCaOG:b-bwo5h2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L1094)):*
> ```Go
> // The locking order is:
> // LeaseManager.mu > tableState.mu > tableNameCache.mu > tableVersionState.mu
> type LeaseManager struct {
> ```

this should be renamed now I think

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
TFTR! I really appreciate you taking a look at this. Thanks

---

Review status: 0 of 7 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/sql/lease.go, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuNAaH4Vcmp0NCv9EM:-KpwV1cQl7IL0hwuUEXp:b-aylbcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L95)):*
> fmt.Sprintf("trying to incRefcount on released lease: %+v", s)
I've added a panic to this code to panic if it sees !s.leased. Removed the "can be used" part.

---

*[pkg/sql/lease.go, line 359 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuOX_HUmYGZMgAKdO2:-KpwW0wA-8hbXWMNTmSz:b-fmb9nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L359)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/now/modTime

And I think we should use `txn.Timestamp()`. In case of serializable txns (which this is), they'll be the same, but `Timestamp` more is what we generally use to indicate the timestamp at which a txn will commit.
</blockquote></details>

there is no txn.Timestamp()

---

*[pkg/sql/lease.go, line 529 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuPMskd5mk2AChjA3h:-KpwXozWFFoRRjICD04X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L529)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this comment should be updated I think
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 547 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuQeH3hp7h_4DLr-K8:-KpwXsCjvYouC-dIOTqW:b-wlgmto) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L547)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm sorry, I don't understand this function. It's supposed to return "the latest version" but it uses a passed-in txn, and sometimes it doesn't go to the database at all. Is the "latest" part just wrong?
</blockquote></details>

We still use the users transaction to acquire leases which is fixed by the next PR in the pipeline. We still assume everywhere in the code that the users transaction is capable of creating a lease at the latest version (we know this is not true if you are using an old transaction). To be fixed in the next PR.

---

*[pkg/sql/lease.go, line 552 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuPRXDrdPJVSS0rbKZ:-KpwWpqwfzJo8gmYMEo0:bbv3045) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L552)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

here `OrigTimestamp` is good (depending on what exactly this method is supposed to do, which I'm not sure), as we probably care about the the transaction's read timestamp
</blockquote></details>

Yeah, in the next PR this is is going to be passed in instead of the entire txn.

---

*[pkg/sql/lease.go, line 554 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuSWyVCO85GrMxrDmX:-KpwYW7i5kenzyE3TwYF:blmwdly) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L554)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

as I was commenting on `hasSomeLifeLeft()`, I'm not sure what it's supposed to do, but I suspect that this will have to change: I think there has to be a case where we refresh a lease async if it's about to expire, but we also go ahead and use it
</blockquote></details>

FYI we do not acquire leases asynchronously. This is a performance issue that will hit us one day (99.9% tail latency). Filed away #17227 

---

*[pkg/sql/lease.go, line 560 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv-7wSmtugH2cJFKHF:-KpwZhaM_vKlgSEN_cob:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L560)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if we have this `findForTimestamp
</blockquote></details>

?

---

*[pkg/sql/lease.go, line 591 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvWGEU9QhTHWFe3z3f:-KpwbwFDMg2HMyuSTTiO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L591)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please comment that the refcount is incremented
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 592 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvarPiF5KTD41LW9fa:-Kpx6lzqPK0WrlZV6Nu-:b6xccz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L592)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

so this doesn't lease, right? Please explain what assumption this has - that a more recent version than timestamp exists in `t.active`?
</blockquote></details>

Yes. I've added a comment explaining the assumptions it makes and what it does

---

*[pkg/sql/lease.go, line 598 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv0CMmH5scpAnXX6zB:-Kpx1DO6TJ_saATIcOOa:bmss50r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L598)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't think we need the loop... Once you go to the store, you're guaranteed to get the right descriptor, aren't you?
</blockquote></details>

I've take the loop out because it was getting hard to explain why we needed it. I've changed the implementation quite a bit.

---

*[pkg/sql/lease.go, line 601 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvaHGU5teOirkmtdds:-Kpx1TOHQUuy8AnRWv2D:b-oldohv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L601)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I found this hard to understand. The `timestamp.Less` are hard to read. I think an ascii diagram might help, for example.
</blockquote></details>

Agreed! Do let me know if it is easier to understand now.

---

*[pkg/sql/lease.go, line 614 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv_LKmZsejaH1FhWpF:-KpwfDCFwm7cFk_8iD5A:bwwhm7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L614)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if this is exclusive, let's please document that where `expiration` is defined
</blockquote></details>

This is documented above the definition of expiration

---

*[pkg/sql/lease.go, line 639 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv_UqjRae_mAcpUah0:-Kpx42kUjRW5c_CrydUn:b-4rubt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L639)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

"valid for the expiration time" sounds very weird. Let's please explain it some more. 
</blockquote></details>

Added some more comments.

---

*[pkg/sql/lease.go, line 640 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv2W1vCqkDJKv_1_Z-:-Kpx2zTThGGf1-aaO5BQ:b-8rzf3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L640)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this should be a method on the `LeaseStore` (which should be renamed) and return a naked descriptor, otherwise it's awkward that this uses the `LeaseStore.db`. And also, if it stays on `tableState`, I think it's weird that it doesn't increment the refcount.
</blockquote></details>

I've moved it to be a method of LeaseStore (not renamed for now). 

---

*[pkg/sql/lease.go, line 640 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvVgp9RIHfi7iH0uGq:-Kpx9IFCsKYofBXmwAJp:b-ksjogl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L640)):*
> replica.requestCanProceed()
Cool! I've added a comment discussing this. I'm not returning a special error code for now, but when I do change the code to be used by AS OF SYSTEM TIMESTAMP I'll explore returning a better error.

---

*[pkg/sql/lease.go, line 648 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv3PIRaS03LHnM7PG6:-Kpx4oDh7wCXyGqPEvHa:bdfiw3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L648)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it'd be :100:  if there was a virtual table for descriptors and we'd do this with a sql time-travel SELECT :)
</blockquote></details>

We'll eventually like to have time travel queries use the LeaseManager (to be renamed) that can also hand out descriptors from the past. I'd hate to add a cyclical dependency here.

---

*[pkg/sql/lease.go, line 751 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpvWs6LFDT2ly_su26V:-Kpx60qJlMEgwumQhKHS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L751)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

removed-removes

Maybe `removeInactive...` is a better name.
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 1094 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv2CGNwcnUfsztCaOF:-Kpx62AYWKoLl0qgWixD:b-ibqwfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L1094)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this should be renamed now I think
</blockquote></details>

I think I'd like to rename quite a few names here. I'll do that in a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-:-KpxBsfnfzB2fQ1V7odw:b426gf5)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Looking good! A few more comments.

---

Review status: 0 of 7 files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease.go, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuNAaH4Vcmp0NCv9EM:-KpzRegy_unVAPHoCQ_P:b-n0a00u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L95)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

> fmt.Sprintf("trying to incRefcount on released lease: %+v", s)
I've added a panic to this code to panic if it sees !s.leased. Removed the "can be used" part.
</blockquote></details>

ok, but then I'd rename to `leaseHasSomeLifeLeft` and document the assertion

---

*[pkg/sql/lease.go, line 359 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuOX_HUmYGZMgAKdO2:-KpzS5eLBAZLmG4IznIb:b-avgbd9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L359)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

there is no txn.Timestamp()
</blockquote></details>

You can add it or access the proto directly

https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/engine/enginepb/mvcc.proto#L58

---

*[pkg/sql/lease.go, line 560 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv-7wSmtugH2cJFKHF:-KpzU-6UP4oEvAMPqSri:b-mtikhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L560)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

?
</blockquote></details>

nvm :)

---

*[pkg/sql/lease.go, line 96 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzUHodDpwcm-5-DNmY:-KpzUHodDpwcm-5-DNmZ:b-l8sui3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L96)):*
> ```Go
> // hasSomeLifeLeft returns true if the lease has at least a minimum of
> // lifetime left until expiration.
> func (s *tableVersionState) hasSomeLifeLeft(timestamp hlc.Timestamp) bool {
> ```

what bothers me now is that, if this method is specific to leases, then I don't see the need for passing in `timestamp` (particularly, a transaction's timestamp). The method should use a clock and look at the current time.
The point of the method is not to tell whether a particular transaction can use a lease or not - that can be simply decided just like it is decided for non-leased `tableVersionStates` (by looking at `expiration`). The point is to decide if we should refresh or not for the benefit of future transactions (and those transactions will presumably have timestamps around the current clock).

That's why I was commenting at the call point about refreshing the lease asynchronously: it's not that I was interested in the performance optimization (although we should definitely do that too), but because I wanted the code to make it clear that the `MinLeaseDuration` has nothing to do with whether the lease can be used or not.
And now that I think about it... if what we do when a lease is about to expire is acquire a new one, and we don't do that asynchronously, then what's the point of doing it preemptively at all (before the lease is expired). Instead of the current query paying the penalty for acquiring the lease, we could let the next one pay. So I'd get rid of this method and `MinLeaseDuration` until we do it async.

---

*[pkg/sql/lease.go, line 454 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzYVicy8_TdSHKjtzx:-KpzYVicy8_TdSHKjtzy:b-ccr7wm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L454)):*
> ```Go
> 		}
> 		if !tableDesc.ModificationTime.Less(prevTimestamp) {
> 			return errors.Errorf("unable to read table = %d with expiration = %s", id, expiration)
> ```

can this condition happen (given that reading below the GC gives you a different error)? If not, I'd add something to the error message to make it sound more unexpected and scary :)

---

*[pkg/sql/lease.go, line 646 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzZjstvDnvvXekWcLO:-KpzZjstvDnvvXekWcLP:b-n9ly33) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L646)):*
> ```Go
> 				return table, nil
> 			}
> 			// table has a valid ModificationTime but an invalid
> ```

I'd replace the first two sentences with "we need a version after data[i], but before data[i+1]."

---

*[pkg/sql/lease.go, line 1066 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzaLeoX7R2YhZglVTu:-KpzaLeoX7R2YhZglVTv:b-u4ez79) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L1066)):*
> ```Go
> 	}
> 
> 	if !table.hasSomeLifeLeft(timestamp) {
> ```

here too I think we should remove this check

---

*[pkg/sql/table.go, line 268 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpzanuf4V2qtGsFqOzz:-Kpzanuf4V2qtGsFqP--:byv3f8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/table.go#L268)):*
> ```Go
> }
> 
> func (tc *TableCollection) maybeChangeTimestamp(ctx context.Context, txn *client.Txn) {
> ```

please add a comment to the `timestamp` member and to this method :)

---

*[pkg/sql/table.go, line 270 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzbZ3Ypk7OD9GaBdLq:-KpzbZ3Ypk7OD9GaBdLr:b-j0rd2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb7a47f1229782e6937158587316c3906d269d75/pkg/sql/table.go#L270)):*
> ```Go
> func (tc *TableCollection) maybeChangeTimestamp(ctx context.Context, txn *client.Txn) {
> 	if tc.timestamp != (hlc.Timestamp{}) && tc.timestamp != txn.OrigTimestamp() {
> 		txn.ResetDeadline()
> ```

I'm not sure why this was here before and I'm not sure if it's ok to remove it. But I'd like to know :). What's the deal?
What happens to a txn's deadline on txn restarts?

---

*[pkg/sql/table.go, line 371 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzbAy8VHYT7FXGDpV4:-KpzbAy8VHYT7FXGDpV5:b9g29gb) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb7a47f1229782e6937158587316c3906d269d75/pkg/sql/table.go#L371)):*
> ```Go
> 	// If the table we just acquired expires before the txn's deadline, reduce
> 	// the deadline.
> 	txn.UpdateDeadlineMaybe(expiration)
> ```

how come this went away? Do we still update the txn's deadline based on descriptor expiration times somewhere? (cause we should, right?)

Btw, let me pass the opportunity to pass over some related stuff :)
Peter and Tobi added the following TODO and put my name on it because of meanness.
https://github.com/cockroachdb/cockroach/blob/8ff5ff97df139fa5958e15a2fd5ffa65e09b49ff/pkg/internal/client/txn.go#L617
I think they're mostly complaining about the comment above that not being clear - it says that the deadline exceeded condition cannot happen, and yet we have some code to deal with it. I think the deadline exceeded condition _can_ happen for SNAPSHOT transactions that have been pushed but don't require a restart.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-:-KpzlrM3p6SjJhKkmfGz:bl0kwno)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
TFTR!

---

Review status: 0 of 5 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/sql/lease.go, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuNAaH4Vcmp0NCv9EM:-Kq0XoNi0iDykU2A-IxB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L95)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ok, but then I'd rename to `leaseHasSomeLifeLeft` and document the assertion
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 359 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuOX_HUmYGZMgAKdO2:-Kq0qJ_7hetBVZPmKCvZ:b-lpb2f1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L359)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

You can add it or access the proto directly

https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/engine/enginepb/mvcc.proto#L58
</blockquote></details>

I don't think I want to do that because this is protected by mutex and I'm not sure what access is safe and if it were to change how our use of it will be affected. I don't think we care about the actual commit timestamp considering we're still imposing a deadline on the txn for the SNAPSHOT ISOLATION case. 

---

*[pkg/sql/lease.go, line 560 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv-7wSmtugH2cJFKHF:-Kq0rkMNhSF1Am_wPDxI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L560)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nvm :)
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 96 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzUHodDpwcm-5-DNmY:-Kq0pBh2bkkTHT4WX_1P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L96)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what bothers me now is that, if this method is specific to leases, then I don't see the need for passing in `timestamp` (particularly, a transaction's timestamp). The method should use a clock and look at the current time.
The point of the method is not to tell whether a particular transaction can use a lease or not - that can be simply decided just like it is decided for non-leased `tableVersionStates` (by looking at `expiration`). The point is to decide if we should refresh or not for the benefit of future transactions (and those transactions will presumably have timestamps around the current clock).

That's why I was commenting at the call point about refreshing the lease asynchronously: it's not that I was interested in the performance optimization (although we should definitely do that too), but because I wanted the code to make it clear that the `MinLeaseDuration` has nothing to do with whether the lease can be used or not.
And now that I think about it... if what we do when a lease is about to expire is acquire a new one, and we don't do that asynchronously, then what's the point of doing it preemptively at all (before the lease is expired). Instead of the current query paying the penalty for acquiring the lease, we could let the next one pay. So I'd get rid of this method and `MinLeaseDuration` until we do it async.
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 454 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzYVicy8_TdSHKjtzx:-Kq0rZKfclmsSCEDRlKw:b-248pyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L454)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can this condition happen (given that reading below the GC gives you a different error)? If not, I'd add something to the error message to make it sound more unexpected and scary :)
</blockquote></details>

This cannot happen. I've marked it as an internal error

---

*[pkg/sql/lease.go, line 646 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzZjstvDnvvXekWcLO:-Kq0s99da9wrCXrAOvMX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L646)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd replace the first two sentences with "we need a version after data[i], but before data[i+1]."
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 1066 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzaLeoX7R2YhZglVTu:-Kq0sH-LTbiIB1SIyjx-:b-rmrsf6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L1066)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

here too I think we should remove this check
</blockquote></details>

I've changed it to use the new hasExpired()

---

*[pkg/sql/table.go, line 268 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpzanuf4V2qtGsFqOzz:-Kq0xhmOm5dhnlz-lq-8:b-pds0lv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/table.go#L268)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please add a comment to the `timestamp` member and to this method :)
</blockquote></details>

I've renamed this method. 

---

*[pkg/sql/table.go, line 270 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzbZ3Ypk7OD9GaBdLq:-Kq0xkcfmFe-HakpRcEJ:b41djw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb7a47f1229782e6937158587316c3906d269d75/pkg/sql/table.go#L270)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm not sure why this was here before and I'm not sure if it's ok to remove it. But I'd like to know :). What's the deal?
What happens to a txn's deadline on txn restarts?
</blockquote></details>

I've put it back for the SNAPSHOT ISOLATION case.

---

*[pkg/sql/table.go, line 371 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpzbAy8VHYT7FXGDpV4:-Kq0y2FV7WWjzj_IiwT0:b-6332xv) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb7a47f1229782e6937158587316c3906d269d75/pkg/sql/table.go#L371)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how come this went away? Do we still update the txn's deadline based on descriptor expiration times somewhere? (cause we should, right?)

Btw, let me pass the opportunity to pass over some related stuff :)
Peter and Tobi added the following TODO and put my name on it because of meanness.
https://github.com/cockroachdb/cockroach/blob/8ff5ff97df139fa5958e15a2fd5ffa65e09b49ff/pkg/internal/client/txn.go#L617
I think they're mostly complaining about the comment above that not being clear - it says that the deadline exceeded condition cannot happen, and yet we have some code to deal with it. I think the deadline exceeded condition _can_ happen for SNAPSHOT transactions that have been pushed but don't require a restart.
</blockquote></details>

put back

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-:-Kq1-lmj0x4LpPWE8cru:b-814g8k)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Do let me know if you have any more comments. Thanks!

---

Review status: 0 of 5 files reviewed at latest revision, 20 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-:-Kq10Ybo-0EW12FK_H3G:bleob81)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I've added a small commit to this PR that I don't have an easy way to test, but I do think it makes sense.

---

Review status: 0 of 5 files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-:-Kq3HpOPGnfN5rdhYUih:b6zmoqk)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

To test the last commit, if I understand it correctly, can't you start a txn, create a table, force a retry, select from the table. Without the patch, I'm not sure what would have happened to the select, but with the patch it should say that the table doesn't exist (which I guess is different than what would have happened before?).

---

Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/lease.go, line 359 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuOX_HUmYGZMgAKdO2:-Kq3jIY1BZhARphU0-HD:b-2ig04q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L359)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I don't think I want to do that because this is protected by mutex and I'm not sure what access is safe and if it were to change how our use of it will be affected. I don't think we care about the actual commit timestamp considering we're still imposing a deadline on the txn for the SNAPSHOT ISOLATION case. 
</blockquote></details>

well you can have an accessor that takes the mutex and then looks in the proto, just like the `OrigTimestamp` accessor does. Here, any access is safe as nobody's sharing the proto yet - you're not doing any operations with the transaction yet.

Anyway, I now no longer care if we use OrigTimestamp or Timestamp here; I think in either case there should be a comment like:
```
// We need to set ModificationTime to the transaction's commit timestamp. Since this is a SERIALZIABLE transaction, that will be OrigTimestamp.
```

---

*[pkg/sql/lease.go, line 648 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv3PIRaS03LHnM7PG6:-Kq3lJzatt8uzmSfMenf:b-nbcbu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L648)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

We'll eventually like to have time travel queries use the LeaseManager (to be renamed) that can also hand out descriptors from the past. I'd hate to add a cyclical dependency here.
</blockquote></details>

Well but I don't think there's any risk of cyclical dependencies if you're reading from a virtual table. Their descriptors don't come from the database. Or do they?

---

*[pkg/sql/lease.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kq3iSTxzk1nFiXlWQeQ:-Kq3iSTxzk1nFiXlWQeR:b4l7dtp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L56)):*
> ```Go
> 	// MinLeaseDuration is the minimum duration a lease will have remaining upon
> 	// acquisition. Exported for testing purposes only.
> 	MinLeaseDuration = time.Minute
> ```

good riddance!

---

*[pkg/sql/lease.go, line 91 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kq3igd3vQYGjZ42xvav:-Kq3igd3vQYGjZ42xvaw:b5etzeb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f87f69ef1f060f53f87b9ebe7117810715a0f9f/pkg/sql/lease.go#L91)):*
> ```Go
> }
> 
> // hasExpired checks if the table has expired for the timestamp.
> ```

nit: s/has expired for the timestamp/is too old to be used (by a txn operating) at the given timestamp

---

*[pkg/sql/lease.go, line 1027 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kq3lnZkuaaM4iFFGjRX:-Kq3lnZkuaaM4iFFGjRY:b-el7ntg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f87f69ef1f060f53f87b9ebe7117810715a0f9f/pkg/sql/lease.go#L1027)):*
> ```Go
> 
> // Resolves a (database ID, table name) to the table descriptor's ID. Returns
> // a valid lease for the table with that name, if the name had been previously
> ```

stale comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-:-Kq3nci5bBuknMCGC8ti:bey0shp)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: ensure the uncommitted descriptor cache is reset properly

When a transaction containing a DDL statement restarts, the
uncommitted cache populated through an earlier attempt of the
transaction is reset.

Also ensure that the SetSystemConfigTrigger() is called when
flushing the old names of a table.

vivekmenezes commented:
I've added a test for the second commit. It took me a while to figure out which schema change would trigger the error.

---

Review status: 0 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/lease.go, line 359 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-KpuOX_HUmYGZMgAKdO2:-Kq4MvxbJSD5JncubydM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L359)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well you can have an accessor that takes the mutex and then looks in the proto, just like the `OrigTimestamp` accessor does. Here, any access is safe as nobody's sharing the proto yet - you're not doing any operations with the transaction yet.

Anyway, I now no longer care if we use OrigTimestamp or Timestamp here; I think in either case there should be a comment like:
```
// We need to set ModificationTime to the transaction's commit timestamp. Since this is a SERIALZIABLE transaction, that will be OrigTimestamp.
```
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 648 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kpv3PIRaS03LHnM7PG6:-Kq4irW8jUsGlyHHoC1H:bwsen0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3d74551c44d15edb1d9a12fc8dc00480babadc3/pkg/sql/lease.go#L648)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Well but I don't think there's any risk of cyclical dependencies if you're reading from a virtual table. Their descriptors don't come from the database. Or do they?
</blockquote></details>

I know very little about them :-)

---

*[pkg/sql/lease.go, line 56 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kq3iSTxzk1nFiXlWQeQ:-Kq4icUIU7FNEDo-C3CZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d761dc3b8e938090f9ff62942ee7beaebacc34dc/pkg/sql/lease.go#L56)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

good riddance!
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 91 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kq3igd3vQYGjZ42xvav:-Kq4MG2UmO5LPzCF14ve:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f87f69ef1f060f53f87b9ebe7117810715a0f9f/pkg/sql/lease.go#L91)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: s/has expired for the timestamp/is too old to be used (by a txn operating) at the given timestamp
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 1027 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-Kq3lnZkuaaM4iFFGjRX:-Kq4iaejooiazrCzvhSR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f87f69ef1f060f53f87b9ebe7117810715a0f9f/pkg/sql/lease.go#L1027)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

stale comment?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16842#-:-Kq62xpLQnS65yfrPOAS:b-u56miv)*
<!-- Sent from Reviewable.io -->
